### PR TITLE
Element Unary Op

### LINKED ISSUE #1056: Update ElementUnary operator

### PR DESCRIPTION
**Pull request by reyna-abhyankar**
**Description of changes:**

Element Unary operator

**Related Issues:**

Linked Issues:
- Issue #1316 opened by this PR (not fixed)

Issues closed by this PR:
- Closes #1056

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)

### CODE REVIEW

**Review by lockshaw — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: all files reviewed, 5 unresolved discussions (waiting on @reyna-abhyankar)

___
*[`lib/kernels/include/kernels/element_unary_kernels.h` line 13 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMeXhSD9aIlxItVeRJ:-NnMeXhSD9aIlxItVeRK:b-30dg8t) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/include/kernels/element_unary_kernels.h#L13)):*
> ```objc
> 
> struct ElementUnaryPerDeviceState {
>   PerDeviceFFHandle handle;
> ```

I don't remember--should this actually be stored as part of the per device state or accessed using the task spec interface as some kind or argument?
___
*[`lib/kernels/include/kernels/element_unary_kernels.h` line 19 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMeksvDp5L7CRMmHG8:-NnMeksvDp5L7CRMmHG9:b-pl8x2r) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/include/kernels/element_unary_kernels.h#L19)):*
> ```objc
>   OperatorType op_type;
>   DataType data_type;
>   float scalar;
> ```

These are part of the op attrs, not the per device state, right?
___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 66 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMfkYsCOdJFd5YZVQ0:-NnMfkYt2_4eaWWgr39o:bzb37ot) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/src/cuda/element_unary_kernels.cu#L66)):*
> ```Cuda
>         break;
>       default:
>         assert(false);
> ```

It would be good to at least start to add an interface for doing more sophisticated error reporting (i.e., having messages) than just `assert`ing out, even if the mechanism for actually handling and showing the error messages doesn't exist yet. This probable shouldn't be done as part of this PR, but it would be good to create an issue to begin tracking this.
___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 72 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMf4L32QpAmkf6UZBK:-NnMf4L32QpAmkf6UZBL:bbkl1yd) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/src/cuda/element_unary_kernels.cu#L72)):*
> ```Cuda
>     checkCUDNN(
>         cudnnSetTensorDescriptorFromArrayShape(inputTensor, input_shape));
>     // input_shape == output_shape
> ```

What is the purpose of this? I'm not sure what this comment is referring to
___
*[`lib/op-attrs/include/op-attrs/ops/element_unary.h` line 24 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMfLxjD5P_bb3zVzv9:-NnMfLxjD5P_bb3zVzvA:b-c7pyax) ([raw file](https://github.com/flexflow/FlexFlow/blob/29406463888a64ef0b62d9103bc470e5f4d16404/lib/op-attrs/include/op-attrs/ops/element_unary.h#L24)):*
> ```objc
> FF_VISITABLE_STRUCT(ElementUnaryAttrs, op);
> CHECK_VALID_OP_ATTR(ElementUnaryAttrs);
> ```

I don't think this should be removed?

**Review by reyna-abhyankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: all files reviewed, 5 unresolved discussions (waiting on @lockshaw)

___
*[`lib/kernels/include/kernels/element_unary_kernels.h` line 13 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMeXhSD9aIlxItVeRJ:-NoOGoLw3GnxKvPhyIkG:bd7cpnf) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/include/kernels/element_unary_kernels.h#L13)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

I don't remember--should this actually be stored as part of the per device state or accessed using the task spec interface as some kind or argument?
</blockquote></details>

Currently, it's kind of ... both? We get handle from argument accessor, and then initialize the device state with it. I think this makes sense, since the handle is needed by the kernel.
___
*[`lib/kernels/include/kernels/element_unary_kernels.h` line 19 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMeksvDp5L7CRMmHG8:-NoOHRDdAABZM_WLf_ZX:b31wjty) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/include/kernels/element_unary_kernels.h#L19)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

These are part of the op attrs, not the per device state, right?
</blockquote></details>

I think it's ok to subsume it in per device state since it kind of streamlines the kernel code. But either is ok, if its attrs then it needs to be passed into the kernel.
___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 66 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMfkYsCOdJFd5YZVQ0:-NoOHmc2E6Ekn3_jdC9j:bqog7zb) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/src/cuda/element_unary_kernels.cu#L66)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

It would be good to at least start to add an interface for doing more sophisticated error reporting (i.e., having messages) than just `assert`ing out, even if the mechanism for actually handling and showing the error messages doesn't exist yet. This probable shouldn't be done as part of this PR, but it would be good to create an issue to begin tracking this.
</blockquote></details>

Agree.
___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 72 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMf4L32QpAmkf6UZBK:-NoOHqwt5nJY6MCYeThS:bqy74n3) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/src/cuda/element_unary_kernels.cu#L72)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

What is the purpose of this? I'm not sure what this comment is referring to
</blockquote></details>

I'm actually not sure why the comment is there either, I think it's pretty obvious that for element unary the input shape == output shape. I think I just wanted to preserve the comment from before about input domain == output domain. But honestly don't think we need it.
___
*[`lib/op-attrs/include/op-attrs/ops/element_unary.h` line 24 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMfLxjD5P_bb3zVzv9:-NoOI7v81DUUQvlneA9m:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/29406463888a64ef0b62d9103bc470e5f4d16404/lib/op-attrs/include/op-attrs/ops/element_unary.h#L24)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

I don't think this should be removed?
</blockquote></details>

Done.

**Review by lockshaw — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: all files reviewed, 3 unresolved discussions (waiting on @reyna-abhyankar)

___
*[`lib/kernels/include/kernels/element_unary_kernels.h` line 13 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMeXhSD9aIlxItVeRJ:-NoOZRA59O9lLIHegVWk:bu9nyw2) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/include/kernels/element_unary_kernels.h#L13)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Currently, it's kind of ... both? We get handle from argument accessor, and then initialize the device state with it. I think this makes sense, since the handle is needed by the kernel.
</blockquote></details>

In that case it seems best to remove it from the `PerDeviceState` and then only pass it in as necessary. Like, if we already have one mechanism to access the `handle`, then having a second one likely opens the possibility to bugs/divergence between the two methods and makes it harder for beginners to reason about which to do. Thoughts?
___
*[`lib/kernels/include/kernels/element_unary_kernels.h` line 19 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMeksvDp5L7CRMmHG8:-NoOZmde6BLQf2M6aSRT:b-agv77k) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/include/kernels/element_unary_kernels.h#L19)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

I think it's ok to subsume it in per device state since it kind of streamlines the kernel code. But either is ok, if its attrs then it needs to be passed into the kernel.
</blockquote></details>

I'd argue that anything that can be passed as an argument should be, and that the device specific feature should only be used where necessary because it's behavior is a bit more complex than the nice value semantics of arguments. Also, it's nicer to just have a single mechanism even if it leads to slightly longer code--see the comment above 
___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 66 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMfkYsCOdJFd5YZVQ0:-NoOZCE589xn5vksAC0t:b-995jdc) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/src/cuda/element_unary_kernels.cu#L66)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Agree.
</blockquote></details>

Can you create an issue, make sure it's added to the project, and then add a link to it here?

**Review by lockshaw — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r3, 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: all files reviewed, 3 unresolved discussions (waiting on @reyna-abhyankar)

*[a discussion](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NozWUOY9UaGLSaCJVxg:-NozWUOY9UaGLSaCJVxh:b-o41jit) (no related file):*
Reminder of the counterargument for removing `ElementScalarUnaryAttrs`: removing it allows many more invalid states (operators that don't use scalars to have scalars, operators that need scalars to not have them) that create more edge cases around hashing, equality, etc. that need to be detected. In addition many of these issues can only be detected at runtime, whereas having two different `Attr` types allows all of these to be prevented at compile-time.

___
___
*[`lib/kernels/src/hip/element_unary_kernels.cpp` line 184 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NozVmIUCyXLqLfKE2fg:-NozVmIUCyXLqLfKE2fh:undefined) ([raw file](https://github.com/flexflow/FlexFlow/blob/39caeefdfe97dcbadb79d00b4a7d946e71e04d6c/lib/kernels/src/hip/element_unary_kernels.cpp#L184)):*
> ```cpp
>                                       m,
>                                       attrs,
>                                       hanlde,
> ```





_Suggestion:_
```cpp
handle
```

**Review by reyna-abhyankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: 8 of 14 files reviewed, 3 unresolved discussions (waiting on @lockshaw)

*[a discussion](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NozWUOY9UaGLSaCJVxg:-Nq54UWZ0DV43qtIyhuR:bkjgq1q) (no related file):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

Reminder of the counterargument for removing `ElementScalarUnaryAttrs`: removing it allows many more invalid states (operators that don't use scalars to have scalars, operators that need scalars to not have them) that create more edge cases around hashing, equality, etc. that need to be detected. In addition many of these issues can only be detected at runtime, whereas having two different `Attr` types allows all of these to be prevented at compile-time.
</blockquote></details>

I've added support for this via inheritance. I think this works and we can keep the code in the rest of the element unary files non-redundant

___

**Review by reyna-abhyankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: 2 of 17 files reviewed, 2 unresolved discussions (waiting on @lockshaw)

___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 66 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMfkYsCOdJFd5YZVQ0:-NrOL42z8RbnV8HX7Pa2:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/src/cuda/element_unary_kernels.cu#L66)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

Can you create an issue, make sure it's added to the project, and then add a link to it here?
</blockquote></details>

Done.
___
*[`lib/op-attrs/include/op-attrs/ops/element_unary.h` line 20 at r6](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nqm9Jyg7X44LkoabtYB:-NrOLA858KjGfmr9X-xY:b-wcy8do) ([raw file](https://github.com/flexflow/FlexFlow/blob/2a567d71d1c8bbc9437481376c04568e1f74af82/lib/op-attrs/include/op-attrs/ops/element_unary.h#L20)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

I don't think this does what you want it to. Inheritance in C++ is done by extension, so `ElementScalarUnaryAttrs` is effectively defined here as 
```
struct ElementScalarUnaryAttrs {
  req<Op> op_type;
  req<optional<float>> scalar;
  req<Op> op_type;
  req<float> scalar;
};
```
which doesn't seem like what you want. I think what you want is more along the lines of 
```
struct ElementUnaryAttrs {
  req<Op> op_type;
};

struct ElementScalarUnaryAttrs : public ElementUnaryAttrs {
  req<float> scalar;
};
```
Either of these approaches has the downside that I no longer have a type that denotes "a unary op attrs *without* a scalar", as now `ElementUnaryAttrs` represents "a unary op attrs with or without a scalar" and `ElementScalarUnaryAttrs` represents "a unary op attrs with a scalar". If you instead have two different and unrelated structs, then you're at least able to represent all three with `ElementScalarUnaryAttrs`, `ElementUnaryAttrs`, and `variant<ElementUnaryAttrs, ElementScalarUnaryAttrs>`. Of course dealing with the `variant` is a bit annoying at times, and so if you *really* think the reduction in code is worthwhile, I think the following would probably work (though I'm not sure the conceptual complexity is worth it--is subtyping really worth it here?)
```
struct GenericElementUnaryAttrs {
  req<Op> op_type;
};

struct NonscalarElementUnaryAttrs : public GenericElementUnaryAttrs { };
struct ScalarElementUnaryAttrs : public GenericElementUnaryAttrs {
  req<float> scalar;
};
```
and you'd have to check whether or not visitable works as expected with this inheritance structure (I think it will, but I'm not 100% certain). I'd also want a concrete example of the case you're worried about with the two-structs-and-a-variant solution to be sure that this more complex solution would actually solve it.

That said, I 100% get the overall impulse. Unfortunately, subtyping in C++ just kinda sucks. There are places (like the graph library) where we're making increased use of the combination of inheritance and visitable, but even then only in cases where the benefits of the additional complexity are clear and concrete--it's also easier to start with the two-structs-and-a-variant solution and move up to this solution if the additional expressiveness is really needed, but I think it's harder to remove the expressiveness of the subtyping if we realize it's more trouble than it's worth as existing code would likely rely  on the behavior.
</blockquote></details>

I changed it to the following and I use `variant<x,y>` as well when needed



_Code snippet:_
```objc
struct ElementUnaryAttrs {
  req<Op> op_type;
};

struct ElementScalarUnaryAttrs {
  req<Op> op_type;
  req<float> scalar;
};
```

**Review by lockshaw — CHANGES_REQUESTED** [review]
> Reviewed 15 of 15 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: all files reviewed, 4 unresolved discussions (waiting on @reyna-abhyankar)

___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 66 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMfkYsCOdJFd5YZVQ0:-Nre0ZjI70kdWIVPTbSv:bi9gue1) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/src/cuda/element_unary_kernels.cu#L66)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Not seeing link? What PR number is it?
___
*[`lib/op-attrs/include/op-attrs/ops/element_unary.h` line 18 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre0hCa-bbH-nI-rzWw:-Nre0hCa-bbH-nI-rzWx:b-928w1m) ([raw file](https://github.com/flexflow/FlexFlow/blob/c85c8dc5fbbec59931468d8de9a7487b21fcfc6d/lib/op-attrs/include/op-attrs/ops/element_unary.h#L18)):*
> ```objc
> 
> struct ElementScalarUnaryAttrs {
>   req<Op> op_type;
> ```

`req` only needed on the last field
___
*[`lib/runtime/src/ops/element_unary.cc` line 27 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre-Q0t-pJzeawSqjzJ:-Nre-Q0t-pJzeawSqjzK:b93h277) ([raw file](https://github.com/flexflow/FlexFlow/blob/c85c8dc5fbbec59931468d8de9a7487b21fcfc6d/lib/runtime/src/ops/element_unary.cc#L27)):*
> ```cpp
> 
> /* ElementUnary */
> OpTaskInvocation init(ElementUnaryUnifiedAttrs const &attrs) {
> ```

This seems good. My only question is if the right overload`init`, `forward`, etc. will be picked up when called on one of the individual types of attrs. If not, it might be good to add trivial wrappers that forward to the variant version. If not, LGTM. 
___
*[`lib/substitutions/include/substitutions/get_attribute.h` line 28 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre0JrC6b3eptGonf5b:-Nre0JrC6b3eptGonf5c:b-hyoy0j) ([raw file](https://github.com/flexflow/FlexFlow/blob/49d2ff7af912f3fd083ccd277d700a3e5bb8a4ab/lib/substitutions/include/substitutions/get_attribute.h#L28)):*
> ```objc
> optional<OperatorAttributeValue> get_attribute(DropoutAttrs const &p,
>                                                OperatorAttributeKey);
> optional<OperatorAttributeValue> get_attribute(ElementBinaryAttrs const &p,
> ```

Why was this removed?

**Review by reyna-abhyankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: all files reviewed, 4 unresolved discussions (waiting on @lockshaw)

___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 66 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-NnMfkYsCOdJFd5YZVQ0:-NrkpfZI0L0lYnnl30Tt:b-334nil) ([raw file](https://github.com/flexflow/FlexFlow/blob/d13e7101e97e4c6e440a1531fb2158d9d6f169a5/lib/kernels/src/cuda/element_unary_kernels.cu#L66)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

Not seeing link? What PR number is it?
</blockquote></details>

I accidentally hit `Done` when we were showing Abishek how to use reviewable LOL. Here it is #1316, i'll link in the PR
___
*[`lib/op-attrs/include/op-attrs/ops/element_unary.h` line 18 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre0hCa-bbH-nI-rzWw:-NrkqzZV3QlUAYHRyHGC:b-q5zwvj) ([raw file](https://github.com/flexflow/FlexFlow/blob/c85c8dc5fbbec59931468d8de9a7487b21fcfc6d/lib/op-attrs/include/op-attrs/ops/element_unary.h#L18)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

`req` only needed on the last field
</blockquote></details>

You mean just `req<float> scalar;`? How come this is the case?
___
*[`lib/runtime/src/ops/element_unary.cc` line 27 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre-Q0t-pJzeawSqjzJ:-Nrkq-xi6Qgc7pN_Q3rO:b-58i4z2) ([raw file](https://github.com/flexflow/FlexFlow/blob/c85c8dc5fbbec59931468d8de9a7487b21fcfc6d/lib/runtime/src/ops/element_unary.cc#L27)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

This seems good. My only question is if the right overload`init`, `forward`, etc. will be picked up when called on one of the individual types of attrs. If not, it might be good to add trivial wrappers that forward to the variant version. If not, LGTM.
</blockquote></details>

I think since attrs is a black box for the task invocations, it won't matter (i.e. the correct one will always be passed to the kernel call and then the kernel function will disambiguate it. 
___
*[`lib/substitutions/include/substitutions/get_attribute.h` line 28 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre0JrC6b3eptGonf5b:-NrkqjZd6mQ3JXQdAaXJ:b-2jjrj7) ([raw file](https://github.com/flexflow/FlexFlow/blob/49d2ff7af912f3fd083ccd277d700a3e5bb8a4ab/lib/substitutions/include/substitutions/get_attribute.h#L28)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

Why was this removed?
</blockquote></details>

`ElementBinary` and `ElementUnary` are already present on lines 22 and 24, for some reason they were duplicated here.

**Review by lockshaw — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: all files reviewed, 2 unresolved discussions (waiting on @reyna-abhyankar)

___
*[`lib/op-attrs/include/op-attrs/ops/element_unary.h` line 18 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre0hCa-bbH-nI-rzWw:-Nrlb33l16Tzb22PvxSR:b-mgt92u) ([raw file](https://github.com/flexflow/FlexFlow/blob/c85c8dc5fbbec59931468d8de9a7487b21fcfc6d/lib/op-attrs/include/op-attrs/ops/element_unary.h#L18)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

You mean just `req<float> scalar;`? How come this is the case?
</blockquote></details>

Since C++ does not support named arguments, if the last member of the struct has to be passed to the constructor then every member of the struct has to be passed to the constructor, which is what we are looking for. There's nothing *wrong* with putting `req` on other fields, but it's unnecessary complication, and for consistency I just decided on the "req only on last field rule"
___
*[`lib/runtime/src/ops/element_unary.cc` line 27 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre-Q0t-pJzeawSqjzJ:-Nrlblo_48mDWUMjxtCQ:bb5qs8j) ([raw file](https://github.com/flexflow/FlexFlow/blob/c85c8dc5fbbec59931468d8de9a7487b21fcfc6d/lib/runtime/src/ops/element_unary.cc#L27)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

I think since attrs is a black box for the task invocations, it won't matter (i.e. the correct one will always be passed to the kernel call and then the kernel function will disambiguate it.
</blockquote></details>

I mean in local backing during the iteration over all of the nodes in the PCG calling `forward`--one of those nodes could be an `ElementUnaryAttrs` and I want to make sure that gets properly resolved to `forward(ElementUnaryUnifiedAttrs const &)`

File: lib/op-attrs/include/op-attrs/ops/element_unary.h [hunk 1]
```diff
@@ -8,20 +8,19 @@
...
+struct ElementUnaryAttrs {
+  req<Op> op_type;
+};
+FF_VISITABLE_STRUCT(ElementUnaryAttrs, op_type);
+CHECK_VALID_OP_ATTR(ElementUnaryAttrs);
+
 struct ElementScalarUnaryAttrs {
-  req<Op> op;
-  /* bool inplace; */
+  req<Op> op_type;
```
Comment (reyna-abhyankar) [inline lib/op-attrs/include/op-attrs/ops/element_unary.h:18]:
```suggestion
Op op_type;
```

File: lib/op-attrs/include/op-attrs/ops/element_unary.h [hunk 2]
```diff
@@ -8,20 +8,19 @@
...
+struct ElementUnaryAttrs {
+  req<Op> op_type;
+};
+FF_VISITABLE_STRUCT(ElementUnaryAttrs, op_type);
+CHECK_VALID_OP_ATTR(ElementUnaryAttrs);
+
 struct ElementScalarUnaryAttrs {
-  req<Op> op;
-  /* bool inplace; */
+Op op_type;
```
Comment (reyna-abhyankar) [inline lib/op-attrs/include/op-attrs/ops/element_unary.h:18]:
```suggestion
    Op op_type;
```

**Review by reyna-abhyankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1257)* status: 16 of 17 files reviewed, 3 unresolved discussions (waiting on @lockshaw)

___
*[`lib/op-attrs/include/op-attrs/ops/element_unary.h` line 18 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre0hCa-bbH-nI-rzWw:-NsLULGeBm5j0Mw6ILXw:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/c85c8dc5fbbec59931468d8de9a7487b21fcfc6d/lib/op-attrs/include/op-attrs/ops/element_unary.h#L18)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

Since C++ does not support named arguments, if the last member of the struct has to be passed to the constructor then every member of the struct has to be passed to the constructor, which is what we are looking for. There's nothing *wrong* with putting `req` on other fields, but it's unnecessary complication, and for consistency I just decided on the "req only on last field rule"
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/element_unary.cc` line 27 at r7](https://reviewable.io/reviews/flexflow/FlexFlow/1257#-Nre-Q0t-pJzeawSqjzJ:-NsLUgAKC8dsyN9iMA6W:b-f1yi19) ([raw file](https://github.com/flexflow/FlexFlow/blob/c85c8dc5fbbec59931468d8de9a7487b21fcfc6d/lib/runtime/src/ops/element_unary.cc#L27)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

I mean in local backing during the iteration over all of the nodes in the PCG calling `forward`--one of those nodes could be an `ElementUnaryAttrs` and I want to make sure that gets properly resolved to `forward(ElementUnaryUnifiedAttrs const &)`
</blockquote></details>

Just tested this in a simple case offline if a caller invokes the function with either of the variant types, and it works.

**Pull request merged**